File: client-android/lib/src/main/kotlin/stasis/client_android/lib/discovery/http/HttpServiceDiscoveryClient.kt
package stasis.client_android.lib.discovery.http

import stasis.client_android.lib.api.clients.internal.ClientExtensions
import stasis.client_android.lib.discovery.ServiceDiscoveryClient
import stasis.client_android.lib.discovery.ServiceDiscoveryResult
import stasis.client_android.lib.security.HttpCredentials
import stasis.client_android.lib.utils.AsyncOps
import stasis.client_android.lib.utils.Try

class HttpServiceDiscoveryClient(
    apiUrl: String,
    override val credentials: suspend () -> HttpCredentials,
    override val attributes: ServiceDiscoveryClient.Attributes
) : ServiceDiscoveryClient, ClientExtensions() {
    override val retryConfig: AsyncOps.RetryConfig = AsyncOps.RetryConfig.Default

    private val server: String = apiUrl.trimEnd { it == '/' }

    override suspend fun latest(isInitialRequest: Boolean): Try<ServiceDiscoveryResult> =
        jsonRequest<ServiceDiscoveryResult> { builder ->
            builder
                .url("$server/v1/discovery/provide")
                .post(attributes.asServiceDiscoveryRequest(isInitialRequest).toBody())
        }
}

File: client-ios/lib/Sources/StasisClientLib/Discovery/Http/HttpServiceDiscoveryClient.swift
import Foundation

public struct HttpServiceDiscoveryClient: ServiceDiscoveryClient, ApiClient {
    public let server: String
    public let attributes: any ServiceDiscoveryClientAttributes
    public let http: HttpClient

    public init(
        apiUrl: String,
        credentialsProvider: any CredentialsProvider,
        attributes: any ServiceDiscoveryClientAttributes,
        retryConfig: RetryConfig = .default
    ) {
        self.server = apiUrl.trimmedTrailingSlash
        self.attributes = attributes
        self.http = HttpClient(
            credentialsProvider: credentialsProvider,
            retryConfig: retryConfig
        )
    }

    init(apiUrl: String, attributes: any ServiceDiscoveryClientAttributes, http: HttpClient) {
        self.server = apiUrl.trimmedTrailingSlash
        self.attributes = attributes
        self.http = http
    }

    public func latest(isInitialRequest: Bool) async throws -> ServiceDiscoveryResult {
        var request = URLRequest(url: URL(string: "\(server)/v1/discovery/provide")!)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try encodeBody(attributes.asServiceDiscoveryRequest(isInitialRequest: isInitialRequest))
        return try await jsonRequest(request)
    }
}
